runtime/src/main/java/io/quarkiverse/logging/cloudwatch/LoggingCloudWatchHandler.java

```diff
@@ -97,49 +99,35 @@ public void close() throws SecurityException {
 
         @Override
         public void run() {
-            while (true) {
-                synchronized (eventBuffer) {
-                    events = new ArrayList<>(eventBuffer);
-                    eventBuffer.clear();
-                }
-
-                if (events.size() > 0) {
-                    PutLogEventsRequest request = new PutLogEventsRequest();
-                    request.setLogEvents(events);
-                    request.setLogGroupName(logGroupName);
-                    request.setLogStreamName(logStreamName);
+            synchronized (eventBuffer) {
+                events = new ArrayList<>(eventBuffer);
+                eventBuffer.clear();
+            }
+            boolean workingSequenceToken = false;
+            if (events.size() > 0) {
+                PutLogEventsRequest request = new PutLogEventsRequest();
+                request.setLogEvents(events);
+                request.setLogGroupName(logGroupName);
+                request.setLogStreamName(logStreamName);
+
+                while (!workingSequenceToken) {
                     request.setSequenceToken(sequenceToken);
-                    // Do the call and get the next token
 
                     try {
                         sequenceToken = awsLogs.putLogEvents(request).getNextSequenceToken();
+                        workingSequenceToken = true;
                     } catch (InvalidSequenceTokenException e) {
                         String exceptionMessage = e.getMessage();
                         LOGGER.info("exception message: " + exceptionMessage);
-                        String validSequenceToken = extractValidSequenceToken(exceptionMessage);
-                        sequenceToken = validSequenceToken;
-                        LOGGER.info("valid sequence token: " + validSequenceToken);
-                        LOGGER.info("actual sequence token: " + sequenceToken);
-
-                        PutLogEventsRequest newRequest = new PutLogEventsRequest();
-                        newRequest.setLogEvents(events);
-                        newRequest.setLogGroupName(logGroupName);
-                        newRequest.setLogStreamName(logStreamName);
-                        newRequest.setSequenceToken(validSequenceToken);
-
-                        sequenceToken = awsLogs.putLogEvents(newRequest).getNextSequenceToken();
+                        sequenceToken = extractValidSequenceToken(exceptionMessage);
+                        LOGGER.info("extracted sequence token: " + sequenceToken);
+                        workingSequenceToken = false;
                     }
                 }
-
-                if (done) {
-                    return;
```

Comment:
As we discussed already, this part of this old code doesn't look good. Logging shouldn't be shut down when `Handler#flush()` is called. That could be the reason why CW logging stopped working a few times on prod.

Reply:
Indeed. Could be.